### PR TITLE
statefulset: use contextual logging

### PR DESCRIPTION
**Pull request by 249043822**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
Switch the statefulset controller code to use contextual and structured logging.
#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/enhancements/issues/3077

#### Does this PR introduce a user-facing change?

```release-note
Migrated the StatefulSet controller (within `kube-controller-manager) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

### CODE REVIEW

File: pkg/controller/statefulset/stateful_set.go [hunk 1]
```diff
@@ -149,8 +161,9 @@ func (ssc *StatefulSetController) Run(ctx context.Context, workers int) {
 
 	defer ssc.queue.ShutDown()
 
-	klog.Infof("Starting stateful set controller")
-	defer klog.Infof("Shutting down statefulset controller")
+	logger := klog.LoggerWithName(klog.FromContext(ctx), "StatefulSet")
```
Comment (pravarag) [inline pkg/controller/statefulset/stateful_set.go:164]: Is there a way to avoid defining multiple `logger` and passing it everywhere but can be defined at the beginning of `statefulset` controller itself?

File: cmd/kube-controller-manager/app/apps.go [hunk 2]
```diff
@@ -50,6 +50,7 @@ func startDaemonSetController(ctx context.Context, controllerContext ControllerC
 
 func startStatefulSetController(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {
 	go statefulset.NewStatefulSetController(
+		ctx,
```
Comment (pohly) [inline cmd/kube-controller-manager/app/apps.go:53]: Let's add a name for the controller here:

```suggestion
		klog.NewContext(ctx, klog.LoggerWithName(klog.FromContext(ctx), "statefulset"))
```

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 3]
```diff
@@ -207,15 +207,16 @@ func (spc *StatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.
 // ClaimsMatchRetentionPolicy returns false if the PVCs for pod are not consistent with set's PVC deletion policy.
 // An error is returned if something is not consistent. This is expected if the pod is being otherwise updated,
 // but a problem otherwise (see usage of this method in UpdateStatefulPod).
-func (spc *StatefulPodControl) ClaimsMatchRetentionPolicy(set *apps.StatefulSet, pod *v1.Pod) (bool, error) {
+func (spc *StatefulPodControl) ClaimsMatchRetentionPolicy(ctx context.Context, set *apps.StatefulSet, pod *v1.Pod) (bool, error) {
+	logger := klog.FromContext(ctx)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_pod_control.go:211]: I think this is one of those cases where the logger is only needed in one unlikely (?) code path, therefore it is better to not look it up here but instead down below if (and only if) needed.

File: pkg/controller/statefulset/stateful_pod_control_test.go [hunk 4]
```diff
@@ -24,8 +24,10 @@ import (
 	"testing"
 	"time"
 
+	"k8s.io/klog/v2/ktesting"
+
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_pod_control_test.go:28]: Remove this blank line.

File: pkg/controller/statefulset/stateful_set.go [hunk 5]
```diff
@@ -78,13 +78,15 @@ type StatefulSetController struct {
...
 func NewStatefulSetController(
+	ctx context.Context,
 	podInformer coreinformers.PodInformer,
 	setInformer appsinformers.StatefulSetInformer,
 	pvcInformer coreinformers.PersistentVolumeClaimInformer,
 	revInformer appsinformers.ControllerRevisionInformer,
 	kubeClient clientset.Interface,
 ) *StatefulSetController {
 	eventBroadcaster := record.NewBroadcaster()
+	logger := klog.LoggerWithName(klog.FromContext(ctx), "StatefulSet")
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:89]: Let's set the name in `cmd/kube-controller-manager/app/core.go` instead. When this code is called in a unit tests we don't need the extra name because there is only one controller.

Comment (pohly) [reply pkg/controller/statefulset/stateful_set.go:89]: Also note that lowercase "statefuleset" might be the preferred name (@pravarag used it below and I think I have seen something similar elsewhere).

File: pkg/controller/statefulset/stateful_set_test.go [hunk 6]
```diff
@@ -311,7 +341,11 @@ func TestStatefulSetControllerAddPod(t *testing.T) {
 }
 
 func TestStatefulSetControllerAddPodOrphan(t *testing.T) {
-	ssc, _, om, _ := newFakeStatefulSetController()
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set_test.go:345]: Why the `context.WithCancel`?

File: pkg/controller/statefulset/stateful_set_test.go [hunk 7]
```diff
@@ -311,7 +341,11 @@ func TestStatefulSetControllerAddPod(t *testing.T) {
...
+	logger := klog.FromContext(ctx)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set_test.go:346]: Instead of taking the context and looking up the logger you can write above:
```
logger, ctx := ktesting.NewTestContext(t)
```

File: pkg/controller/statefulset/stateful_set_test.go [hunk 8]
```diff
@@ -325,17 +359,21 @@ func TestStatefulSetControllerAddPodOrphan(t *testing.T) {
...
 		t.Errorf("queue.Len() = %v, want %v", got, want)
 	}
 }
 
 func TestStatefulSetControllerAddPodNoSet(t *testing.T) {
-	ssc, _, _, _ := newFakeStatefulSetController()
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	logger := klog.FromContext(ctx)
+	defer cancel()
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set_test.go:372]: See above... also applies below.

File: pkg/controller/statefulset/stateful_set.go (hunk 1 above)
Comment (pohly) [reply pkg/controller/statefulset/stateful_set.go:164]: We could put a logger into `ssc *StatefulSetController`, but I am not a fan of that: it then becomes much too easy to write code which should log something that got passed down the call chain and doesn't because it uses this `ssc.logger` instance.

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/statefulset/stateful_set.go [hunk 9]
```diff
@@ -437,7 +444,7 @@ func (ssc *StatefulSetController) worker(ctx context.Context) {
 func (ssc *StatefulSetController) sync(ctx context.Context, key string) error {
 	startTime := time.Now()
 	defer func() {
-		klog.V(4).Infof("Finished syncing statefulset %q (%v)", key, time.Since(startTime))
+		ssc.logger.V(4).Info("Finished syncing statefulset", "statefulSet", key, "time", time.Since(startTime))
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:447]: This is why I don't like objects that contain a logger: this function gets a context which might contain a logger which is different from ssc.logger, but it uses the cached logger instead.

Can you remove ssc.logger and just pass through a logger instance where needed (i.e. for the handler functions)?

File: pkg/controller/statefulset/stateful_set_test.go [hunk 10]
```diff
@@ -940,7 +967,7 @@ func getPodAtOrdinal(pods []*v1.Pod, ordinal int) *v1.Pod {
 	return pods[ordinal]
 }
 
-func scaleUpStatefulSetController(set *apps.StatefulSet, ssc *StatefulSetController, spc *StatefulPodControl, om *fakeObjectManager) error {
+func scaleUpStatefulSetController(logger klog.Logger, set *apps.StatefulSet, ssc *StatefulSetController, spc *StatefulPodControl, om *fakeObjectManager) error {
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set_test.go:970]: This doesn't need a logger?

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/statefulset/stateful_set.go (hunk 1 above)
Comment (pohly) [reply pkg/controller/statefulset/stateful_set.go]: And I have seen exactly that error being made elsewhere...

File: pkg/controller/statefulset/stateful_set.go [hunk 11]
```diff
@@ -98,21 +100,26 @@ func NewStatefulSetController(
...
 		),
-		pvcListerSynced: pvcInformer.Informer().HasSynced,
-		revListerSynced: revInformer.Informer().HasSynced,
-		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "statefulset"),
-		podControl:      controller.RealPodControl{KubeClient: kubeClient, Recorder: recorder},
-
+		pvcListerSynced:  pvcInformer.Informer().HasSynced,
+		revListerSynced:  revInformer.Informer().HasSynced,
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "statefulset"),
+		podControl:       controller.RealPodControl{KubeClient: kubeClient, Recorder: recorder},
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:106]: This looks like an unrelated change that just happened because one blank line got removed, forcing a reformatting.

You don't need to if this is the only thing that needs to be changed, but it would be cleaner to remove this change.

File: pkg/controller/statefulset/stateful_set.go [hunk 12]
```diff
@@ -124,7 +131,7 @@ func NewStatefulSetController(
 				oldPS := old.(*apps.StatefulSet)
 				curPS := cur.(*apps.StatefulSet)
 				if oldPS.Status.Replicas != curPS.Status.Replicas {
-					klog.V(4).Infof("Observed updated replica count for StatefulSet: %v, %d->%d", curPS.Name, oldPS.Status.Replicas, curPS.Status.Replicas)
+					logger.V(4).Info("Observed updated replica count for StatefulSet", "statefulSet", curPS.Name, "oldReplicas", oldPS.Status.Replicas, "newReplicas", curPS.Status.Replicas)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:134]:
```suggestion
					logger.V(4).Info("Observed updated replica count for StatefulSet", "statefulSet", klog.KObj(curPS), "oldReplicas", oldPS.Status.Replicas, "newReplicas", curPS.Status.Replicas)
```

File: pkg/controller/statefulset/stateful_set.go [hunk 13]
```diff
@@ -180,7 +188,7 @@ func (ssc *StatefulSetController) addPod(obj interface{}) {
 		if set == nil {
 			return
 		}
-		klog.V(4).Infof("Pod %s created, labels: %+v", pod.Name, pod.Labels)
+		logger.V(4).Info("Pod created with labels", "pod", pod.Name, "labels", pod.Labels)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:191]:
```suggestion
		logger.V(4).Info("Pod created with labels", "pod", klog.KObj(pod), "labels", pod.Labels)
```

Same below.

File: pkg/controller/statefulset/stateful_set.go [hunk 14]
```diff
@@ -436,8 +444,9 @@ func (ssc *StatefulSetController) worker(ctx context.Context) {
 // sync syncs the given statefulset.
 func (ssc *StatefulSetController) sync(ctx context.Context, key string) error {
 	startTime := time.Now()
+	logger := klog.FromContext(ctx)
 	defer func() {
-		klog.V(4).Infof("Finished syncing statefulset %q (%v)", key, time.Since(startTime))
+		logger.V(4).Info("Finished syncing statefulset", "statefulSet", key, "time", time.Since(startTime))
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:449]: Above you are using "statefulSet" for an object, here's it's the key. Let's use "key" as key name here.

File: pkg/controller/statefulset/stateful_set.go [hunk 15]
```diff
@@ -446,7 +455,7 @@ func (ssc *StatefulSetController) sync(ctx context.Context, key string) error {
 	}
 	set, err := ssc.setLister.StatefulSets(namespace).Get(name)
 	if errors.IsNotFound(err) {
-		klog.Infof("StatefulSet has been deleted %v", key)
+		logger.Info("StatefulSet has been deleted", "statefulSet", key)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:458]: Also here.

File: pkg/controller/statefulset/stateful_set.go [hunk 16]
```diff
@@ -475,14 +484,15 @@ func (ssc *StatefulSetController) sync(ctx context.Context, key string) error {
...
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Syncing StatefulSet with pods", "statefulSet", klog.KRef(set.Namespace, set.Name), "pods", len(pods))
 	var status *apps.StatefulSetStatus
 	var err error
 	status, err = ssc.control.UpdateStatefulSet(ctx, set, pods)
 	if err != nil {
 		return err
 	}
-	klog.V(4).Infof("Successfully synced StatefulSet %s/%s successful", set.Namespace, set.Name)
+	logger.V(4).Info("Successfully synced StatefulSet", "statefulSet", klog.KRef(set.Namespace, set.Name))
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:495]:
```suggestion
	logger.V(4).Info("Successfully synced StatefulSet", "statefulSet", klog.KObj(set))
```

File: pkg/controller/statefulset/stateful_set_test.go [hunk 17]
```diff
@@ -50,8 +52,9 @@ func alwaysReady() bool { return true }
 
 func TestStatefulSetControllerCreates(t *testing.T) {
 	set := newStatefulSet(3)
-	ssc, spc, om, _ := newFakeStatefulSetController(set)
-	if err := scaleUpStatefulSetController(set, ssc, spc, om); err != nil {
+	_, ctx := ktesting.NewTestContext(t)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set_test.go:55]:
```suggestion
	logger, ctx := ktesting.NewTestContext(t)
```


Then below you don't need klog.FromContext.

Same in other tests.

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 18]
```diff
@@ -207,15 +207,15 @@ func (spc *StatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.
...
+func (spc *StatefulPodControl) ClaimsMatchRetentionPolicy(ctx context.Context, set *apps.StatefulSet, pod *v1.Pod) (bool, error) {
 	ordinal := getOrdinal(pod)
 	templates := set.Spec.VolumeClaimTemplates
 	for i := range templates {
 		claimName := getPersistentVolumeClaimName(set, &templates[i], ordinal)
 		claim, err := spc.objectMgr.GetClaim(set.Namespace, claimName)
 		switch {
 		case apierrors.IsNotFound(err):
-			klog.V(4).Infof("Expected claim %s missing, continuing to pick up in next iteration", claimName)
+			klog.FromContext(ctx).V(4).Info("Expected claim missing, continuing to pick up in next iteration", "claim", claimName)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_pod_control.go:218]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md mentions "PVC" as good key for a `PersistentVolumeClaim` object.

Please replace ` "claim", claimName` with `"PVC", klog.KRef(..., claimName)` throughout the entire PR. Use `klog.KObj` when there is a claim object.

File: pkg/controller/statefulset/stateful_set_control.go [hunk 19]
```diff
@@ -646,10 +648,9 @@ func updateStatefulSetAfterInvariantEstablished(
 		// delete the Pod if it is healthy and the revision doesnt match the target
 		if getPodRevision(replicas[target]) != updateRevision.Name && !isTerminating(replicas[target]) {
 			// delete the Pod if it is healthy and the revision doesnt match the target
-			klog.V(2).Infof("StatefulSet %s/%s terminating Pod %s for update",
-				set.Namespace,
-				set.Name,
-				replicas[target].Name)
+			logger.V(2).Info("StatefulSet terminating Pod for update",
+				"statefulSet", klog.KObj(set),
+				"pod", replicas[target].Name)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set_control.go:653]: Log a `kobj.KRef` as value for "pod".

**Review by pohly — CHANGES_REQUESTED** [review]

**Comment by pohly** [comment]
Was rebasing onto a new base revision necessary? If not, then please avoid it. https://github.com/kubernetes/kubernetes/compare/268b70da64c0591b44653a76f4f5a38908a5a903..7dfc748bfc5c17d6b2a23560780ca6a8691be5bd isn't helpful now.

File: pkg/controller/statefulset/stateful_set.go (hunk 11 above)
Comment (pohly) [reply pkg/controller/statefulset/stateful_set.go:106]: This is still open.

File: pkg/controller/statefulset/stateful_set.go [hunk 20]
```diff
@@ -225,13 +233,13 @@ func (ssc *StatefulSetController) updatePod(old, cur interface{}) {
...
 		}
-		klog.V(4).Infof("Pod %s updated, objectMeta %+v -> %+v.", curPod.Name, oldPod.ObjectMeta, curPod.ObjectMeta)
+		logger.V(4).Info("Pod objectMeta updated", "pod", klog.KObj(curPod), "oldObjectMeta", oldPod.ObjectMeta, "newObjectMeta", curPod.ObjectMeta)
 		ssc.enqueueStatefulSet(set)
 		// TODO: MinReadySeconds in the Pod will generate an Available condition to be added in
 		// the Pod status which in turn will trigger a requeue of the owning replica set thus
 		// having its status updated with the newly available replica.
 		if !podutil.IsPodReady(oldPod) && podutil.IsPodReady(curPod) && set.Spec.MinReadySeconds > 0 {
-			klog.V(2).Infof("StatefulSet %s will be enqueued after %ds for availability check", set.Name, set.Spec.MinReadySeconds)
+			logger.V(2).Info("StatefulSet will be enqueued after minReadySeconds for availability check", "statefulSet", set.Name, "minReadySeconds", set.Spec.MinReadySeconds)
```
Comment (pohly) [inline pkg/controller/statefulset/stateful_set.go:242]:
```suggestion
			logger.V(2).Info("StatefulSet will be enqueued after minReadySeconds for availability check", "statefulSet", klog.KObj(set), "minReadySeconds", set.Spec.MinReadySeconds)
```

**Review by pohly — CHANGES_REQUESTED** [review]

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by sftim** [comment]
The changelog entry I suggest for this:
```markdown
Migrated the StatefulSet controller (within `kube-controller-manager) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

**Comment by pohly** [comment]
@249043822: please add the changelog entry in your PR description.

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 21]
```diff
@@ -207,15 +207,15 @@ func (spc *StatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.
...
+func (spc *StatefulPodControl) ClaimsMatchRetentionPolicy(ctx context.Context, set *apps.StatefulSet, pod *v1.Pod) (bool, error) {
 	ordinal := getOrdinal(pod)
 	templates := set.Spec.VolumeClaimTemplates
 	for i := range templates {
 		claimName := getPersistentVolumeClaimName(set, &templates[i], ordinal)
 		claim, err := spc.objectMgr.GetClaim(set.Namespace, claimName)
 		switch {
 		case apierrors.IsNotFound(err):
-			klog.V(4).Infof("Expected claim %s missing, continuing to pick up in next iteration", claimName)
+			klog.FromContext(ctx).V(4).Info("Expected claim missing, continuing to pick up in next iteration", "PVC", klog.KObj(claim))
```
Comment (obaranov1) [inline pkg/controller/statefulset/stateful_pod_control.go:218]: why do we have here klog.FromContext(ctx).V(4).Info("..."), and not logger.V(4).Info("...")?

**Comment by pohly** [comment]
@249043822: the release note must go into the PR description body, not the title.

    ```release-note
     Migrated the StatefulSet controller (within `kube-controller-manager) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging)
     ```

/title statefulset: use contextual logging

**Comment by 249043822** [comment]
> @249043822: the release note must go into the PR description body, not the title.
> 
> ```
> ```release-note
>  Migrated the StatefulSet controller (within `kube-controller-manager) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging)
>  ```
> ```
> 
> /title statefulset: use contextual logging

done

**Review by pohly — APPROVED** [review]
/lgtm
/assign @soltysh 

For approval.

**Review by soltysh — APPROVED** [review]
/approve

**Comment by pacoxu** [comment]
/lgtm cancel
for verify failure 
```
pkg/controller/statefulset/stateful_pod_control.go:327:55: not enough arguments in call to spc.UpdatePodClaimForRetentionPolicy
	have (*"k8s.io/api/apps/v1".StatefulSet, *"k8s.io/api/core/v1".Pod)
	want (context.Context, *"k8s.io/api/apps/v1".StatefulSet, *"k8s.io/api/core/v1".Pod)
```

It may caused by rebasing.

**Comment by pohly** [comment]
@249043822: please fix the build failures. If they don't occur in your branch, then you have to rebase because there was some other change in the master branch that isn't getting noticed by "git merge" as a conflict.

**Comment by 249043822** [comment]
Done

**Review by pohly — APPROVED** [review]
/lgtm
